Issue 1974

amansr02 opened the pull request:
# I have made things!

<!--
Hi, thanks for submitting a Pull Request. We appreciate it.

Please, fill in all the required information
to make our review and merging processes easier.

Cheers!
-->

## Checklist

<!-- Please check everything that applies: -->

- [x] I have double checked that there are no unrelated changes in this pull request (old patches, accidental config files, etc)
- [x] I have created at least one test case for the changes I have made
- [x] I have updated the documentation for the changes I have made
- [x] I have added my changes to the `CHANGELOG.md`

## Related issues

<!--
Mark what issues this Pull Request closes or references.

Format is:
- Closes #issue-number
- Refs #issue-number

Example. Refs #0
Documentation: https://blog.github.com/2013-05-14-closing-issues-via-pull-requests/
-->

<!--
If you have any feedback, just write it here.

It can be whatever you want!
-->

🙏 Please, if you or your company is finding wemake-python-styleguide valuable, help us sustain the project by sponsoring it transparently on https://opencollective.com/wemake-python-styleguide. As a thank you, your profile/company logo will be added to our main README which receives hundreds of unique visitors per day.

skarzi commented on tests/fixtures/noqa/noqa.py:
```diff
@@ -788,3 +788,24 @@ def get_item():  # noqa: WPS463
 ]
+
+def detect_bare_raise1():
+    """Function to check if the bare raise is detected."""  # noqa: DAR401
```
Let's resolve all docstrings' related violations instead of ignoring them

skarzi commented on wemake_python_styleguide/visitors/ast/keywords.py:
```diff
@@ -71,6 +80,11 @@ def _check_exception_type(self, node: ast.Raise) -> None:
 
+    def _check_bare_raise(self, node: ast.Raise) -> None:
+        parent = walk.get_closest_parent(node, ast.ExceptHandler)
+        if not parent:
```
you also need to check  that `node.exc is None` to ensure that the given `raise` statement is bare

skarzi commented on tests/fixtures/noqa/noqa.py:
```diff
@@ -788,3 +788,24 @@ def get_item():  # noqa: WPS463
    for number in numbers # noqa: WPS361
 ]
+
+def detect_bare_raise1():
```
I think that something like:

```python
def bare_raise():
    raise
```

will be enough

skarzi reviewed (changes requested):
Thanks, that's PR with awesome changes, but it needs some polishing and few additional unit tests that check the newly introduced violation.
You can add a new test file to `tests/test_visitors/test_ast/test_keywords/`.
Let's check e.g. [`tests/test_visitors/test_ast/test_keywords/test_raise.py`](https://github.com/wemake-services/wemake-python-styleguide/blob/master/tests/test_visitors/test_ast/test_keywords/test_raise.py) file with tests related to the `raise` statement (so you need to add quite similar ones).
If you have any questions etc, ping me here :smile:

amansr02 commented:
Hi @skarzi. Thank you for reviewing our changes. They should be fixed now. We have integrated those changes and have added test cases. Flake8 runs perfectly on wemake-python-styleguide project. Do let us know if we need to change other things as well?

skarzi commented on tests/fixtures/noqa/noqa.py:
```diff
@@ -788,3 +788,20 @@ def get_item():  # noqa: WPS463
    for number in numbers # noqa: WPS361
 ]
+
+def detect_bare_raise1():
+    """Function to check if the bare raise is detected."""
+    raise
+
+def detect_bare_raise2_helper():
+    """Function to check if the bare raise is detected."""
+    raise ZeroDivisionError
+
+def detect_bare_raise2():
+    """Function to check if the bare raise is detected."""
+    try:
+        testvar = 1/0  # noqa: WPS344, F841
+    except ZeroDivisionError:
+        detect_bare_raise2()
+        raise ZeroDivisionError
+
```
Let's leave here just one function with bare `raise` like mentioned in https://github.com/wemake-services/wemake-python-styleguide/pull/1998/files#r619652218

skarzi commented on tests/test_visitors/test_ast/test_keywords/test_raise.py:
```diff
@@ -130,3 +131,22 @@ def test_bare_raise(
+    code = """
+    def hello():
+        raise
+    hello()
```
function call is probably unnecessary:

```suggestion
```

skarzi commented on tests/test_visitors/test_ast/test_keywords/test_raise.py:
```diff
@@ -130,3 +131,22 @@ def test_bare_raise(
     assert_errors(visitor, [])
+
+
+def test_bare_raise_no_except(
```
let's add tests (or even better one parametrized test) for following cases:

+ bare `raise` inside `except` block
+ `raise SomeException()` inside `except` block
+ `raise SomeException()`  not nested in `except` block

skarzi commented on wemake_python_styleguide/logic/safe_eval.py:
```diff
@@ -14,7 +14,6 @@ def _convert_num(node: Optional[ast.AST]):
     elif isinstance(node, ast.Name):
         # We return string names as is, see how we return strings:
         return node.id
-    raise ValueError('malformed node or string: {0!r}'.format(node))
```
why you have removed this line? 
Let's revert it

skarzi commented on wemake_python_styleguide/visitors/ast/keywords.py:
```diff
@@ -71,6 +74,11 @@ def _check_exception_type(self, node: ast.Raise) -> None:
             )
 
+    def _check_bare_raise(self, node: ast.Raise) -> None:
+        parent = walk.get_closest_parent(node, ast.ExceptHandler)
+        if not parent and node.exc is None:
+            self.add_violation(BareRaiseViolation(node))
```
NIP: to optimize this code we can first check if the given `node` is bare `raise` statement (it's quite rarely used compared to `raise SomeException()`, so mostly it should evaluate to `False`) and then check its parent:

```suggestion
        if node.exc is None:
            if not walk.get_closest_parent(node, ast.ExceptHandler):
                self.add_violation(BareRaiseViolation(node))
```

skarzi reviewed (changes requested):
great progress, just a few things are worth adding/refactoring, before shipping this feature!

amansr02 commented:
Hi! I have updated my changes and fixed one of the noqa test cases as well to reflect our added changes. Not sure if I was supposed to change the test_noqa function. Do let me know if this change works out! Greatly appreciate the feedback.

skarzi commented on tests/test_visitors/test_ast/test_keywords/test_raise.py:
```diff
@@ -130,3 +131,60 @@ def test_bare_raise(
 
     assert_errors(visitor, [])
+
+
+def test_raise_except(
+    assert_errors,
+    parse_ast_tree,
+    default_options,
+):
+    """Testing that bare `raise` is not allowed without an except block."""
+    code = """
+    def hello():
+        try:
+            x = 1
+        except Exception:
+            raise ValueError('1')
+    """
```
Let's parametrize this test to test at least 2 cases: 

+ `raise SomeException()` inside `except` block
+ bare `raise` inside `except` block

```suggestion

@pytest.mark.parametrize('raise_statement', [
    "raise ValueError('1')",
    'raise',
])
def test_raise_inside_except(
    assert_errors,
    parse_ast_tree,
    default_options,
    raise_statement,
):
    """Ensure ``raise`` statement inside ``except`` block allowed."""
    code = """
        try:
            x = 1
        except Exception:
            {raise_statement}
    """.format(raise_statement=raise_statement)
```

skarzi replied:
You can also create one parametrized test like suggested by Nikita in https://github.com/wemake-services/wemake-python-styleguide/pull/1998#discussion_r619993569 or me in https://github.com/wemake-services/wemake-python-styleguide/pull/1998#discussion_r619823903

skarzi commented on tests/test_visitors/test_ast/test_keywords/test_raise.py:
```diff
@@ -130,3 +131,60 @@ def test_bare_raise(
+    assert_errors(visitor, [])
+
+
+def test_raise_no_except(
```
```suggestion
def test_raise_exception_no_except(
```

sobolevn reviewed (changes requested):
Thanks a lot!

skarzi commented on tests/test_visitors/test_ast/test_keywords/test_raise.py:
```diff
@@ -130,3 +131,60 @@ def test_bare_raise(
+    parse_ast_tree,
+    default_options,
+):
+    """Testing that bare `raise` is not allowed without an except block."""
```
```suggestion
    """Testing that  `raise SomeException()` outside ``except`` block is allowed."""
```

skarzi reviewed (changes requested):
great work, I just have added a few more comments and change requests/suggestions

amansr02 commented:
I parameterized the test functions and now have 1 testcase that is correct and 1 testcase that raises an error in the linter.
Also updated the noqa.py. Thank you for the feedback guys! Really helping me become familiar with the codebase. Any other changes?

skarzi commented on tests/test_visitors/test_ast/test_keywords/test_raise.py:
```diff
@@ -112,21 +121,43 @@ def test_raise_good_errors(
-    """
+    """Testing correct instances of `raise SomeException()` and bare `raise`."""
+    code = None
+    if bare_option:
```
personally, I dislike using `if` statements in tests, because they make tests harder to read and maintain, so I am suggesting to parameterized function with ready to parse code strings, e.g.:

```python
@pytest.mark.parametrize('code', [
    raise_exception_raw.format('ValueError(1)'),
    raise_exception_with_except_raw.format('ValueError(1)'),
    raise_exception_with_except_raw.format(''),
])
def test_correct raise_statement(
    assert_errors,
    parse_ast_tree,
    default_options,
    code,
 ):
     ...
````

skarzi commented on tests/test_visitors/test_ast/test_keywords/test_raise.py:
```diff
@@ -17,6 +18,14 @@ def check_exception_without_call():
     raise {0}
 """
 
+raise_exception_with_except = """
+def check_exception_without_call():
+    try:
+        x = 1
+    except Exception:
+        raise {0}
```
NIP: it's not necessary to wrap `try except` in function definition:

```suggestion
raise_exception_with_except_raw = """
try:
    x = 1
except Exception:
    raise {0}
```

sobolevn commented:
Implemented as `WPS467`
